File: packages/python/infermap/infermap/identity.py
"""InferMap -> Identity Graph bridge.

Helper that writes InferMap's schema-mapping output into the GoldenMatch
Identity Graph as ``IdentityAlias`` rows. When InferMap discovers that
``crm.cust_id`` maps to ``customer_id`` (the canonical entity-id field
on the target schema), each source record's ``crm.cust_id`` value becomes
an alias that resolves to that record's identity.

This is **per-record** alias writing -- InferMap tells us *which column
holds the id of this kind*, we record one row per (record, alias-kind).
Schema-level "this column maps to that column" aliasing without a record
context is intentionally **not** modeled -- the alias table is keyed on
the alias *value*, not the column name.

Sized down on purpose: this module imports goldenmatch.identity lazily so
infermap users without goldenmatch installed keep working. The
``write_aliases_from_mapping`` function is the only public surface.

See issue #206 for the design discussion.
"""
from __future__ import annotations

import logging
from collections.abc import Callable, Iterable
from dataclasses import dataclass
from typing import Any

from infermap.types import MapResult

logger = logging.getLogger(__name__)


@dataclass
class AliasWriteResult:
    """Summary of one ``write_aliases_from_mapping`` invocation."""

    aliases_written: int = 0
    records_processed: int = 0
    mappings_used: int = 0
    skipped_low_confidence: int = 0
    skipped_no_value: int = 0
    skipped_no_entity: int = 0
    skipped_no_kind: int = 0

    def as_dict(self) -> dict[str, int]:
        return {
            "aliases_written": self.aliases_written,
            "records_processed": self.records_processed,
            "mappings_used": self.mappings_used,
            "skipped_low_confidence": self.skipped_low_confidence,
            "skipped_no_value": self.skipped_no_value,
            "skipped_no_entity": self.skipped_no_entity,
            "skipped_no_kind": self.skipped_no_kind,
        }


def _is_alias_kind(target_field: str, alias_kinds: frozenset[str]) -> str | None:
    """Return the alias ``kind`` for a target field name, or None.

    Strict match against ``alias_kinds`` plus a tiny set of common
    aliases (e.g. ``customer_id`` -> ``customer_id``). Case-insensitive.
    """
    norm = target_field.lower().strip()
    if norm in alias_kinds:
        return norm
    return None


def write_aliases_from_mapping(
    mapping: MapResult,
    records: Iterable[dict[str, Any]],
    store: Any,
    entity_id_resolver: Callable[[dict[str, Any]], str | None],
    *,
    source_name: str,
    alias_kinds: frozenset[str] = frozenset({
        "customer_id", "user_id", "account_id", "external_id", "email",
        "phone", "ssn", "tax_id", "ein", "vin", "isbn", "doi",
    }),
    min_confidence: float = 0.85,
    dataset: str | None = None,
) -> AliasWriteResult:
    """Write IdentityAlias rows for each record where InferMap mapped a
    source column to a known alias-kind target column.

    Parameters
    ----------
    mapping:
        InferMap's ``MapResult`` -- typically the result of
        ``infermap.map(source_schema, target_schema)``.
    records:
        Iterable of dicts. Each dict represents one source record;
        keys are source field names.
    store:
        A ``goldenmatch.identity.IdentityStore`` instance. Imported lazily;
        this function will not import goldenmatch if the caller doesn't
        pass one.
    entity_id_resolver:
        ``record -> entity_id | None`` function. Typically pulls the
        record's primary-key value and calls
        ``store.find_entity_by_record(record_id)``. Returning None for a
        record skips alias writing for that row.
    source_name:
        The source name (e.g. ``"crm"``). Used to namespace the alias
        value (``f"{source_name}:{value}"``) so two sources with the
        same id space don't collide.
    alias_kinds:
        Target field names that count as alias-kinds. The default set
        covers the common identifier types; pass an extended set when
        your target schema has domain-specific ids
        (e.g. ``"npi"`` for healthcare).
    min_confidence:
        Drop any mapping below this confidence threshold. 0.85 default
        matches the threshold InferMap considers a "strong" match.
    dataset:
        Optional dataset name flowed onto each ``IdentityAlias`` row.

    Returns
    -------
    AliasWriteResult
        Counters describing what got written and why anything was
        skipped.
    """
    try:
        from goldenmatch.identity import IdentityAlias
    except ImportError as e:
        raise ImportError(
            "write_aliases_from_mapping requires goldenmatch>=1.15.0 "
            "to be installed (provides IdentityAlias).",
        ) from e

    # Pre-compute the usable (source_col, target_kind) tuples from the
    # mapping. Drops low-confidence and non-alias-kind mappings upfront so
    # we don't reread per record.
    usable: list[tuple[str, str]] = []
    for m in mapping.mappings:
        if m.confidence < min_confidence:
            continue
        kind = _is_alias_kind(m.target, alias_kinds)
        if kind is None:
            continue
        usable.append((m.source, kind))

    skipped_low = sum(
        1 for m in mapping.mappings if m.confidence < min_confidence
    )
    summary = AliasWriteResult(
        mappings_used=len(usable),
        skipped_low_confidence=skipped_low,
    )
    if not usable:
        logger.info(
            "No usable alias-kind mappings (>= %.2f confidence). "
            "Available mappings: %s",
            min_confidence,
            [m.target for m in mapping.mappings],
        )
        return summary

    for record in records:
        summary.records_processed += 1
        entity_id = entity_id_resolver(record)
        if entity_id is None:
            summary.skipped_no_entity += 1
            continue

        for source_col, kind in usable:
            value = record.get(source_col)
            if value is None or value == "":
                summary.skipped_no_value += 1
                continue
            alias_value = f"{source_name}:{value}"
            try:
                store.add_alias(
                    IdentityAlias(
                        alias=alias_value,
                        entity_id=entity_id,
                        kind=kind,
                        dataset=dataset,
                    ),
                )
                summary.aliases_written += 1
            except Exception as e:  # noqa: BLE001
                # Don't blow up the whole batch on one bad row -- log and
                # continue. Identity is additive; partial writes are fine.
                logger.warning(
                    "Failed to write alias %s (kind=%s) for entity %s: %s",
                    alias_value, kind, entity_id, e,
                )

    logger.info(
        "InferMap -> Identity: wrote %d aliases across %d records (%d mappings used)",
        summary.aliases_written,
        summary.records_processed,
        summary.mappings_used,
    )
    return summary


__all__ = ["AliasWriteResult", "write_aliases_from_mapping"]

File: packages/typescript/infermap/src/node/identity.ts
/**
 * InferMap -> Identity Graph bridge (node-only).
 *
 * Helper that writes InferMap's schema-mapping output into the GoldenMatch
 * Identity Graph as `IdentityAlias` rows. When InferMap discovers that
 * `crm.cust_id` maps to `customer_id` (the canonical entity-id field on the
 * target schema), each source record's `crm.cust_id` value becomes an alias
 * that resolves to that record's identity.
 *
 * This is **per-record** alias writing -- InferMap tells us *which column
 * holds the id of this kind*, we record one row per (record, alias-kind).
 * Schema-level "this column maps to that column" aliasing without a record
 * context is intentionally **not** modeled -- the alias table is keyed on the
 * alias *value*, not the column name.
 *
 * Node-only: it depends on the `goldenmatch` Identity Graph (`IdentityStore` /
 * `IdentityAlias`). The edge-safe core never imports this module.
 *
 * TS parity with `packages/python/infermap/infermap/identity.py`. The Python
 * sibling lazily imports `goldenmatch.identity` and raises a clean ImportError
 * if it isn't installed; in TS the dependency is declared in package.json so
 * the type-only import never fails at runtime. `store.addAlias` is async on the
 * TS `IdentityStore` interface, so this function is async too and the
 * `entityIdResolver` may return a value or a Promise.
 *
 * See issue #206 for the design discussion.
 */

import type { IdentityAlias, IdentityStore } from "goldenmatch/core";

import type { MapResult } from "../core/types.js";

/** Summary of one `writeAliasesFromMapping` invocation. */
export interface AliasWriteResult {
  aliasesWritten: number;
  recordsProcessed: number;
  mappingsUsed: number;
  skippedLowConfidence: number;
  skippedNoValue: number;
  skippedNoEntity: number;
  skippedNoKind: number;
}

/** Serialize an AliasWriteResult to the Python `as_dict()` snake_case shape. */
export function aliasWriteResultAsDict(
  r: AliasWriteResult,
): Record<string, number> {
  return {
    aliases_written: r.aliasesWritten,
    records_processed: r.recordsProcessed,
    mappings_used: r.mappingsUsed,
    skipped_low_confidence: r.skippedLowConfidence,
    skipped_no_value: r.skippedNoValue,
    skipped_no_entity: r.skippedNoEntity,
    skipped_no_kind: r.skippedNoKind,
  };
}

/**
 * Target field names that count as alias-kinds by default. Mirrors the Python
 * `alias_kinds` default. Pass an extended set when your target schema has
 * domain-specific ids (e.g. `"npi"` for healthcare).
 */
export const DEFAULT_ALIAS_KINDS: ReadonlySet<string> = new Set([
  "customer_id",
  "user_id",
  "account_id",
  "external_id",
  "email",
  "phone",
  "ssn",
  "tax_id",
  "ein",
  "vin",
  "isbn",
  "doi",
]);

/** Default minimum confidence -- matches InferMap's "strong match" threshold. */
export const DEFAULT_MIN_CONFIDENCE = 0.85;

/**
 * Return the alias `kind` for a target field name, or null. Strict
 * case-insensitive match against `aliasKinds`. Mirrors Python `_is_alias_kind`.
 */
function isAliasKind(
  targetField: string,
  aliasKinds: ReadonlySet<string>,
): string | null {
  const norm = targetField.toLowerCase().trim();
  return aliasKinds.has(norm) ? norm : null;
}

export type EntityIdResolver = (
  record: Record<string, unknown>,
) => string | null | undefined | Promise<string | null | undefined>;

export interface WriteAliasesOptions {
  /** Source name (e.g. `"crm"`). Namespaces the alias value (`source:value`). */
  sourceName: string;
  /**
   * Target field names that count as alias-kinds. Defaults to
   * {@link DEFAULT_ALIAS_KINDS}.
   */
  aliasKinds?: ReadonlySet<string> | Iterable<string>;
  /** Drop any mapping below this confidence. Defaults to 0.85. */
  minConfidence?: number;
  /** Optional dataset name flowed onto each `IdentityAlias` row. */
  dataset?: string | null;
  /**
   * Logger invoked when a single `addAlias` call throws. Defaults to
   * `console.warn`. A bad row never aborts the batch (identity is additive).
   */
  onError?: (info: {
    alias: string;
    kind: string;
    entityId: string;
    error: unknown;
  }) => void;
}

/**
 * Write `IdentityAlias` rows for each record where InferMap mapped a source
 * column to a known alias-kind target column.
 *
 * @param mapping InferMap's `MapResult` (e.g. from `map(source, target)`).
 * @param records Iterable of source records (dicts keyed by source field name).
 * @param store A goldenmatch `IdentityStore` instance.
 * @param entityIdResolver `record -> entityId | null` (may be async). Returning
 *   null/undefined skips alias writing for that row.
 * @param options See {@link WriteAliasesOptions}; `sourceName` is required.
 */
export async function writeAliasesFromMapping(
  mapping: MapResult,
  records: Iterable<Record<string, unknown>>,
  store: IdentityStore,
  entityIdResolver: EntityIdResolver,
  options: WriteAliasesOptions,
): Promise<AliasWriteResult> {
  if (store === null || store === undefined || typeof store.addAlias !== "function") {
    throw new Error(
      "writeAliasesFromMapping requires a goldenmatch IdentityStore " +
        "(an object exposing addAlias).",
    );
  }

  const {
    sourceName,
    minConfidence = DEFAULT_MIN_CONFIDENCE,
    dataset = null,
  } = options;
  const aliasKinds: ReadonlySet<string> =
    options.aliasKinds === undefined
      ? DEFAULT_ALIAS_KINDS
      : options.aliasKinds instanceof Set
        ? (options.aliasKinds as ReadonlySet<string>)
        : new Set(options.aliasKinds);
  const onError =
    options.onError ??
    (({ alias, kind, entityId, error }) => {
      // eslint-disable-next-line no-console
      console.warn(
        `Failed to write alias ${alias} (kind=${kind}) for entity ${entityId}: ${String(error)}`,
      );
    });

  // Pre-compute the usable (sourceCol, targetKind) tuples. Drops
  // low-confidence and non-alias-kind mappings upfront so we don't reread
  // per record.
  const usable: Array<[string, string]> = [];
  let skippedLowConfidence = 0;
  for (const m of mapping.mappings) {
    if (m.confidence < minConfidence) {
      skippedLowConfidence += 1;
      continue;
    }
    const kind = isAliasKind(m.target, aliasKinds);
    if (kind === null) continue;
    usable.push([m.source, kind]);
  }

  const summary: AliasWriteResult = {
    aliasesWritten: 0,
    recordsProcessed: 0,
    mappingsUsed: usable.length,
    skippedLowConfidence,
    skippedNoValue: 0,
    skippedNoEntity: 0,
    skippedNoKind: 0,
  };

  if (usable.length === 0) {
    return summary;
  }

  for (const record of records) {
    summary.recordsProcessed += 1;
    const entityId = await entityIdResolver(record);
    if (entityId === null || entityId === undefined) {
      summary.skippedNoEntity += 1;
      continue;
    }

    for (const [sourceCol, kind] of usable) {
      const value = record[sourceCol];
      if (value === null || value === undefined || value === "") {
        summary.skippedNoValue += 1;
        continue;
      }
      const aliasValue = `${sourceName}:${value}`;
      const aliasRow: IdentityAlias = {
        alias: aliasValue,
        entityId,
        kind,
        dataset,
        recordedAt: new Date(),
      };
      try {
        await store.addAlias(aliasRow);
        summary.aliasesWritten += 1;
      } catch (error) {
        // Don't blow up the whole batch on one bad row -- log and continue.
        // Identity is additive; partial writes are fine.
        onError({ alias: aliasValue, kind, entityId, error });
      }
    }
  }

  return summary;
}
